src/record_layout.zig

```diff
@@ -0,0 +1,459 @@
+    }
+
+    fn layoutRegluarField(self: *MscvContext, size_bits: u64, field_align: u32, fld: *Field) void {
+        self.contains_non_bitfield = true;
+        self.ongoing_bitfield = null;
+        self.pointer_align_bits = std.math.max(self.pointer_align_bits, field_align);
+        self.field_align_bits = std.math.max(self.field_align_bits, field_align);
+        const offset_bits = switch (self.is_union) {
+            true => 0,
+            false => std.mem.alignForwardGeneric(u64, self.size_bits, field_align),
+        };
+        self.size_bits = std.math.max(self.size_bits, offset_bits + size_bits);
+        fld.layout = .{
+            .offset_bits = offset_bits,
+            .size_bits = size_bits,
+        };
+    }
+};
+
+pub fn recordLayout(ty: *Type, comp: *const Compilation, parser: ?*const Parser) void {
+    // const mapper = comp.string_interner.getSlowTypeMapper();
+    // const name = mapper.lookup(ty.getRecord().?.name);
+    // std.debug.print("struct {s}\n", .{name});
+    switch (comp.langopts.emulate) {
+        .gcc, .clang => {
+            var context = SysVContext.init(ty, comp, parser);
+            var rec = getMutableRecord(ty).?;
+
+            context.layoutFields(rec, comp);
+
+            context.size_bits = std.mem.alignForwardGeneric(u64, context.size_bits, context.aligned_bits);
+
+            rec.type_layout = TypeLayout{
+                .size_bits = context.size_bits,
+                .field_alignment_bits = context.aligned_bits,
+                .pointer_alignment_bits = context.aligned_bits,
+                .required_alignment_bits = BITS_PER_BYTE,
+            };
+        },
+        .msvc => {
+            var context = MscvContext.init(ty, comp, parser);
+            var rec = getMutableRecord(ty).?;
+
+            for (rec.fields) |*fld, fld_indx| {
+                var field_attrs: ?[]const Attribute = null;
+                if (rec.field_attributes) |attrs| {
+                    field_attrs = attrs[fld_indx];
+                }
+
+                context.layoutField(comp, fld, field_attrs);
+            }
+            rec.type_layout = TypeLayout{
+                .size_bits = context.size_bits,
+                .field_alignment_bits = context.field_align_bits,
+                .pointer_alignment_bits = context.pointer_align_bits,
+                .required_alignment_bits = context.req_align_bits,
+            };
+        },
+    }
+}
+
+pub fn computeLayout(ty: Type, comp: *const Compilation, type_layout: *TypeLayout) void {
+    if (ty.isRecord()) {
+        const rec = ty.getRecord().?;
+        type_layout.* = rec.type_layout;
+    } else {
+        type_layout.size_bits = ty.bitSizeof(comp) orelse 0;
+        type_layout.pointer_alignment_bits = ty.alignof(comp) * BITS_PER_BYTE;
+        type_layout.field_alignment_bits = type_layout.pointer_alignment_bits;
+        type_layout.required_alignment_bits = BITS_PER_BYTE;
+    }
+}
+
+pub fn getMutableRecord(ty: *Type) ?*Type.Record {
```

Comment:
```suggestion
pub fn getMutableRecord(ty: *Type) *Type.Record {
```
This whole function is a bit awkward but it should definitely not return an optional.

Reply:
It's a clone of `getRecord()` that only differs by `const` in the return type. I didn't want to change `getRecord` because it's used in other places.